### PR TITLE
webgpu: Use WGPU poller thread for poll_all_devices

### PR DESCRIPTION
**Pull request by sagudev**
As discussed https://servo.zulipchat.com/#narrow/stream/263398-general/topic/ipc_channel, firefox will also switch to something similar in the future: https://bugzilla.mozilla.org/show_bug.cgi?id=1870699.

In future we could make thread per device, but that would require hashmap for Pollers.

Fastgame still works: https://sagudev.github.io/briefcase/fastgame.html

try run: https://github.com/sagudev/servo/actions/runs/9051091523/job/24867638272

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes in WebGPU CTS

### CODE REVIEW

**Comment by sagudev** [comment]
Timings are more sane now (we actually outperform firefox 🎉 ):
Servo run onSubmittedWorkDone:* test in 130ms

**Comment by sagudev** [comment]
We are still missing poll somewhere because `webgpu:api,operation,compute,basic:large_dispatch:*` has flaky TIMEOUT (unless we poll every loop iteration, this could also be locking race in wgpu).

**Comment by sagudev** [comment]
I think there might be deadlock between poll and submit

**Comment by sagudev** [comment]
Some flakes that do occur but rarely (sometimes they are stable, other times they simply do not happen):
- `CRASH [expected OK] /_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,buffer,mapping:mapAsync,state,destroyed:*`
with:
```
called `Result::unwrap()` on an `Err` value: Destroyed (thread WGPU poller, at components/webgpu/wgpu_thread.rs:112)
```
- `ERROR /_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,buffer,mapping:mapAsync,state,mappingPending:*`
- `TIMEOUT [expected OK] /_webgpu/webgpu/cts.https.html?q=webgpu:api,operation,compute,basic:large_dispatch:*`

File: components/webgpu/poll_thread.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                            // this is mostly useful for stuff that is deferred
+                            // to maintain calls in wgpu (device resource destruction)
+                            poll_all_devices(&global, &mut more_work, false);
+                            while more_work || work.load(Ordering::Acquire) != 0 {
+                                poll_all_devices(&global, &mut more_work, true);
+                            }
+                            std::thread::park(); //TODO: should we use timeout here
+                        }
+                    })
+                    .unwrap(),
```
Comment (gterzian) [inline components/webgpu/poll_thread.rs:79]: expect?

**Review by gterzian — APPROVED** [review]
LGTM, with some nits and perhaps one refactoring.

**Comment by gterzian** [comment]
> (this is actually our fault as we do not handle this situation).

File an issue for this one? Seems like we can catch the destroy error and send an error back to script?

**Comment by sagudev** [comment]
> > (this is actually our fault as we do not handle this situation).
> 
> File an issue for this one? Seems like we can catch the destroy error and send an error back to script?

Done https://github.com/servo/servo/issues/32277

File: components/webgpu/poll_thread.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        self.handle
+            .as_ref()
+            .expect("Pollers thread does not exist!")
+            .thread()
+            .unpark();
+    }
+}
+
+impl Drop for Poller {
+    fn drop(&mut self) {
```
[the remark this answers is not in the document]
Comment (gterzian) [reply components/webgpu/poll_thread.rs:107]: right...

**Review by gterzian — APPROVED** [review]
LGTM

**Pull request merged**